File: everything_parsing/16-lexer-project1/js/minimize.mjs
// Module 14 — DFA minimization.
//
// Two states are EQUIVALENT if the DFA accepts the same continuations from
// either. Merge each equivalence class → the MINIMAL DFA, unique by Myhill-Nerode.
//
//   partitionRefinement — {accepting}|{non-accepting}, split any block whose
//                         members disagree on where a symbol sends them (Moore)
//   tableFilling         — mark distinguishable pairs to a fixed point

import { Dfa } from "./dfa.mjs";
import { build } from "./thompson.mjs";
import { determinize } from "./subset.mjs";

// ── remove unreachable states, keep original names ──

export function trim(d) {
  const reach = new Set([d.start]);
  const q = [d.start];
  while (q.length) {
    const s = q.shift();
    for (const a of d.alphabet) {
      const t = d.step(s, a);
      if (!reach.has(t)) { reach.add(t); q.push(t); }
    }
  }
  if (reach.size === d.states.length) return d;

  let text = "states: " + [...reach].join(" ") + "\n";
  text += "alphabet: " + d.alphabet.join(" ") + "\n";
  text += "start: " + d.start + "\n";
  text += "accept: " + [...reach].filter(s => d.accept.has(s)).join(" ") + "\n";
  for (const s of reach)
    for (const a of d.alphabet)
      text += s + " " + a + " " + d.step(s, a) + "\n";
  return Dfa.parse(text);
}

// ── partition refinement (Moore) ──

export function partitionRefinement(d0) {
  const d = trim(d0);
  const states = d.states;
  const block = new Map();
  for (const s of states) block.set(s, d.accept.has(s) ? 1 : 0);
  let count = new Set(block.values()).size;

  for (;;) {
    const sigId = new Map();
    const next = new Map();
    for (const s of states) {
      const sig = [block.get(s), ...d.alphabet.map(a => block.get(d.step(s, a)))].join(",");
      if (!sigId.has(sig)) sigId.set(sig, sigId.size);
      next.set(s, sigId.get(sig));
    }
    const nc = new Set(next.values()).size;
    for (const [k, v] of next) block.set(k, v);
    if (nc === count) break;
    count = nc;
  }
  return blocksToSets(states, block);
}

// ── table filling ──

export function tableFilling(d0) {
  const d = trim(d0);
  const states = d.states;
  const n = states.length;
  const idx = new Map(states.map((s, i) => [s, i]));
  const marked = Array.from({ length: n }, () => new Array(n).fill(false));

  for (let i = 0; i < n; i++)
    for (let j = i + 1; j < n; j++)
      marked[i][j] = d.accept.has(states[i]) !== d.accept.has(states[j]);

  let changed = true;
  while (changed) {
    changed = false;
    for (let i = 0; i < n; i++)
      for (let j = i + 1; j < n; j++) {
        if (marked[i][j]) continue;
        for (const a of d.alphabet) {
          const pi = idx.get(d.step(states[i], a));
          const pj = idx.get(d.step(states[j], a));
          const lo = Math.min(pi, pj), hi = Math.max(pi, pj);
          if (lo !== hi && marked[lo][hi]) { marked[i][j] = true; changed = true; break; }
        }
      }
  }

  const parent = states.map((_, i) => i);
  const find = x => (parent[x] === x ? x : (parent[x] = find(parent[x])));
  for (let i = 0; i < n; i++)
    for (let j = i + 1; j < n; j++)
      if (!marked[i][j]) parent[find(i)] = find(j);

  const classes = new Map();
  for (let i = 0; i < n; i++) {
    const r = find(i);
    if (!classes.has(r)) classes.set(r, new Set());
    classes.get(r).add(states[i]);
  }
  return [...classes.values()].map(s => [...s].sort());
}

// ── quotient (minimal) DFA ──

export function minimal(d0) {
  const d = trim(d0);
  let classes = partitionRefinement(d).map(s => [...s].sort());
  classes.sort((x, y) => {
    const xs = x.includes(d.start), ys = y.includes(d.start);
    if (xs !== ys) return xs ? -1 : 1;
    return x[0] < y[0] ? -1 : x[0] > y[0] ? 1 : 0;
  });

  const nameOf = new Map();
  classes.forEach((c, i) => c.forEach(s => nameOf.set(s, "M" + i)));

  let text = "states: " + classes.map((_, i) => "M" + i).join(" ") + "\n";
  text += "alphabet: " + d.alphabet.join(" ") + "\n";
  text += "start: " + nameOf.get(d.start) + "\n";
  text += "accept: " + classes.map((c, i) => (d.accept.has(c[0]) ? "M" + i : null)).filter(Boolean).join(" ") + "\n";
  classes.forEach((c, i) => {
    for (const a of d.alphabet) text += "M" + i + " " + a + " " + nameOf.get(d.step(c[0], a)) + "\n";
  });
  return Dfa.parse(text);
}

// ── regex equivalence via minimal-DFA isomorphism ──

export function equivalent(a, b) {
  return isomorphic(minimal(determinize(build(a))), minimal(determinize(build(b))));
}

export function isomorphic(a, b) {
  if (a.states.length !== b.states.length) return false;
  if (a.alphabet.join(",") !== b.alphabet.join(",")) return false;
  const map = new Map([[a.start, b.start]]);
  const q = [[a.start, b.start]];
  while (q.length) {
    const [sa, sb] = q.shift();
    if (a.accept.has(sa) !== b.accept.has(sb)) return false;
    for (const sym of a.alphabet) {
      const ta = a.step(sa, sym), tb = b.step(sb, sym);
      if (!map.has(ta)) { map.set(ta, tb); q.push([ta, tb]); }
      else if (map.get(ta) !== tb) return false;
    }
  }
  return true;
}

// ── helpers ──

function blocksToSets(states, block) {
  const g = new Map();
  for (const s of states) {
    const b = block.get(s);
    if (!g.has(b)) g.set(b, new Set());
    g.get(b).add(s);
  }
  return [...g.values()].map(s => [...s].sort());
}

File: everything_parsing/16-lexer-project1/java/Minimize.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.TreeSet;

/**
 * Module 14 — DFA minimization.
 *
 * Two states are EQUIVALENT if the DFA accepts exactly the same continuation
 * strings from either. Merge every equivalence class into one state → the
 * MINIMAL DFA, which by the Myhill-Nerode theorem is UNIQUE for the language.
 *
 *   partitionRefinement  — start {accepting} | {non-accepting}, keep splitting
 *                          a block whose members disagree on which block some
 *                          symbol sends them to, until stable (Moore's method)
 *   tableFilling         — mark "obviously distinguishable" pairs (one accepting
 *                          one not), then mark a pair if some symbol sends it to
 *                          an already-marked pair, to a fixed point
 *
 * Both yield the same partition. `minimal()` builds the quotient DFA.
 */
public final class Minimize {
    private Minimize() {}

    // ── remove states unreachable from the start (keeps original names) ──

    static Dfa trim(Dfa d) {
        Set<String> reach = new LinkedHashSet<>();
        Deque<String> q = new ArrayDeque<>(List.of(d.start));
        reach.add(d.start);
        while (!q.isEmpty()) {
            String s = q.poll();
            for (String a : d.alphabet) {
                String t = d.step(s, a);
                if (reach.add(t)) q.add(t);
            }
        }
        if (reach.size() == d.states.size()) return d;

        StringBuilder sb = new StringBuilder("states:");
        for (String s : reach) sb.append(" ").append(s);
        sb.append("\nalphabet:");
        for (String a : d.alphabet) sb.append(" ").append(a);
        sb.append("\nstart: ").append(d.start).append("\naccept:");
        for (String s : reach) if (d.accept.contains(s)) sb.append(" ").append(s);
        sb.append("\n");
        for (String s : reach)
            for (String a : d.alphabet)
                sb.append(s).append(" ").append(a).append(" ").append(d.step(s, a)).append("\n");
        return Dfa.parse(sb.toString());
    }

    // ── partition refinement (Moore) ──

    static List<Set<String>> partitionRefinement(Dfa d) {
        d = trim(d);
        List<String> states = new ArrayList<>(d.states);
        List<String> alpha = new ArrayList<>(d.alphabet);

        Map<String, Integer> block = new HashMap<>();
        for (String s : states) block.put(s, d.accept.contains(s) ? 1 : 0);
        int blockCount = distinct(block);

        while (true) {
            Map<List<Integer>, Integer> sigId = new LinkedHashMap<>();
            Map<String, Integer> next = new HashMap<>();
            for (String s : states) {
                List<Integer> sig = new ArrayList<>();
                sig.add(block.get(s));
                for (String a : alpha) sig.add(block.get(d.step(s, a)));
                next.put(s, sigId.computeIfAbsent(sig, k -> sigId.size()));
            }
            int nc = distinct(next);
            block = next;
            if (nc == blockCount) break;   // no block was split this round
            blockCount = nc;
        }
        return blocksToSets(states, block);
    }

    // ── table filling ──

    static List<Set<String>> tableFilling(Dfa d) {
        d = trim(d);
        List<String> states = new ArrayList<>(d.states);
        int n = states.size();
        Map<String, Integer> idx = new HashMap<>();
        for (int i = 0; i < n; i++) idx.put(states.get(i), i);
        boolean[][] marked = new boolean[n][n];

        for (int i = 0; i < n; i++)
            for (int j = i + 1; j < n; j++)
                marked[i][j] = d.accept.contains(states.get(i)) != d.accept.contains(states.get(j));

        boolean changed = true;
        while (changed) {
            changed = false;
            for (int i = 0; i < n; i++)
                for (int j = i + 1; j < n; j++) {
                    if (marked[i][j]) continue;
                    for (String a : d.alphabet) {
                        int pi = idx.get(d.step(states.get(i), a));
                        int pj = idx.get(d.step(states.get(j), a));
                        int lo = Math.min(pi, pj), hi = Math.max(pi, pj);
                        if (lo != hi && marked[lo][hi]) { marked[i][j] = true; changed = true; break; }
                    }
                }
        }

        int[] parent = new int[n];
        for (int i = 0; i < n; i++) parent[i] = i;
        for (int i = 0; i < n; i++)
            for (int j = i + 1; j < n; j++)
                if (!marked[i][j]) union(parent, i, j);

        Map<Integer, Set<String>> classes = new LinkedHashMap<>();
        for (int i = 0; i < n; i++)
            classes.computeIfAbsent(find(parent, i), k -> new TreeSet<>()).add(states.get(i));
        return new ArrayList<>(classes.values());
    }

    // ── build the quotient (minimal) DFA ──

    static Dfa minimal(Dfa d) {
        d = trim(d);
        List<Set<String>> classes = new ArrayList<>(partitionRefinement(d));
        // order: the class with the start state first, then by smallest member
        final Dfa dd = d;
        classes.sort((x, y) -> {
            boolean xs = x.contains(dd.start), ys = y.contains(dd.start);
            if (xs != ys) return xs ? -1 : 1;
            return new TreeSet<>(x).first().compareTo(new TreeSet<>(y).first());
        });

        Map<String, String> nameOf = new HashMap<>();
        for (int i = 0; i < classes.size(); i++)
            for (String s : classes.get(i)) nameOf.put(s, "M" + i);

        StringBuilder sb = new StringBuilder("states:");
        for (int i = 0; i < classes.size(); i++) sb.append(" M").append(i);
        sb.append("\nalphabet:");
        for (String a : d.alphabet) sb.append(" ").append(a);
        sb.append("\nstart: ").append(nameOf.get(d.start)).append("\naccept:");
        for (int i = 0; i < classes.size(); i++) {
            String rep = new TreeSet<>(classes.get(i)).first();
            if (d.accept.contains(rep)) sb.append(" M").append(i);
        }
        sb.append("\n");
        for (int i = 0; i < classes.size(); i++) {
            String rep = new TreeSet<>(classes.get(i)).first();
            for (String a : d.alphabet)
                sb.append("M").append(i).append(" ").append(a).append(" ").append(nameOf.get(d.step(rep, a))).append("\n");
        }
        return Dfa.parse(sb.toString());
    }

    // ── regex equivalence: minimal DFAs are isomorphic iff the languages are equal ──

    static boolean equivalent(Regex a, Regex b) {
        return isomorphic(
            minimal(Subset.determinize(Thompson.build(a))),
            minimal(Subset.determinize(Thompson.build(b))));
    }

    static boolean isomorphic(Dfa a, Dfa b) {
        if (a.states.size() != b.states.size()) return false;
        if (!a.alphabet.equals(b.alphabet)) return false;
        Map<String, String> map = new HashMap<>();
        Deque<String[]> q = new ArrayDeque<>();
        map.put(a.start, b.start);
        q.add(new String[]{a.start, b.start});
        while (!q.isEmpty()) {
            String[] pr = q.poll();
            if (a.accept.contains(pr[0]) != b.accept.contains(pr[1])) return false;
            for (String sym : a.alphabet) {
                String ta = a.step(pr[0], sym), tb = b.step(pr[1], sym);
                String seen = map.get(ta);
                if (seen == null) { map.put(ta, tb); q.add(new String[]{ta, tb}); }
                else if (!seen.equals(tb)) return false;
            }
        }
        return true;
    }

    // ── helpers ──

    private static int distinct(Map<String, Integer> m) { return new HashSet<>(m.values()).size(); }
    private static List<Set<String>> blocksToSets(List<String> states, Map<String, Integer> block) {
        Map<Integer, Set<String>> g = new LinkedHashMap<>();
        for (String s : states) g.computeIfAbsent(block.get(s), k -> new TreeSet<>()).add(s);
        return new ArrayList<>(g.values());
    }
    private static int find(int[] p, int x) { return p[x] == x ? x : (p[x] = find(p, p[x])); }
    private static void union(int[] p, int a, int b) { p[find(p, a)] = find(p, b); }
}
